### DISCUSSION TITLE
Keeping order of received DNS servers

### DISCUSSION THREAD

**Discussion by oldium** [strongswan/strongswan#1732] [Ideas]
The `resolve` plugin keeps the list of received DNS servers, but it sorts them in the array. I think it would be great to keep the received DNS order - same order as the VPN configuration author intended. What do you think?

I have an actual problem with the behaviour of a `resolve` plugin. The VPN configuration is made in a way that the first DNS server can answer internal and also external/internet host queries and the second DNS server is a fallback with only external/internet hosts to at least have a working internet when being connected to VPN and first/main DNS is down. The `resolve` plugin changes order of those DNS, so the internal host name resolution simply does not work.

**Comment by tobiasbrunner** [comment]
Thanks for the report. We can try keeping the order. I've pushed a change to that effect to the _1732-resolve-order_ branch.

> The VPN configuration is made in a way that the first DNS server can answer internal and also external/internet host queries and the second DNS server is a fallback with only external/internet hosts to at least have a working internet when being connected to VPN and first/main DNS is down.

Relying on the order is not ideal in general, as stub resolvers are free to treat all configured DNS servers equally (unless the resolver supports some kind of restrictions, e.g. a list of domains, that could be applied to some of them, but that won't be the case here). Some resolvers might even query all servers e.g. to determine if one of them is significantly faster or for load balancing purposes.

You could perhaps configure a DNS proxy like `dnsmasq` on your VPN server and configure the DNS servers there as you need them (e.g. the internal one restricted to only specific domains) and then hand clients the internal IP of your VPN server as single DNS server address.

**Reply by oldium** [reply]
I am using `dnsmasq` for name resolution. As an input I use `/etc/resolv.conf` file generated either by StrongSwan's `resolve` plugin or let it go through `openresolv` (my config allows to use both approaches).

StrongSwan in my use case binds two VPNs together, one where it acts as a client (here it receives list of DNS servers) and one where it acts as a server (here it offers IP address of the `dnsmasq` instance as a DNS server, the `dnsmasq` instance uses the generated `resolv.conf` file from the first VPN).

**Comment by oldium** [comment]
Great, thanks a lot, I will try it. From the code I see that it is fixed for openresolv/resolvconf case. Maybe it would be good to have it also for /etc/resolv.conf modification case? Now for the /etc/resolv.conf it is in reversed order - it is added one by one to the top of the file.

**Reply by tobiasbrunner** [reply]
> Now for the /etc/resolv.conf it is in reversed order - it is added one by one to the top of the file.

Hm, but that was always the case and isn't affected by the sorting of the servers in the array you were referring to.

**Reply by oldium** [reply]
Yes, it is not a new issue, I just wanted to mention that there is (and was) inconsistency - maybe it would be good to use similar/same approach, i.e. always generate the full DNS list, which is passed to openresolv/resolvconf or written to /etc/resolv.conf (removing all generated entries would need to be done first). So that both openresolv/resolvconf and /etc/resolv.conf would have the same order of DNS servers.

**Reply by tobiasbrunner** [reply]
I see. I've had a look at this and pushed a commit to the branch that changes how DNS servers are installed in `/etc/resolv.conf`. Similar to the `resolvconf` approach, for every installed or removed server, the complete set of available servers is written to the file in order. Afterwards, all previously installed servers (identified by a common suffix) are simply skipped when the contents of the old file are copied. Let me know if that works for you.

**Comment by oldium** [comment]
Unfortunately build failed under alpine:edge Docker image (gcc 13.1.1) - cast to `struct sockaddr*` is missing. I prepared quick-fix in my tree https://github.com/oldium/strongswan/tree/1732-resolve-order

```
dhcp_socket.c: In function 'dhcp_socket_create':
dhcp_socket.c:902:45: error: passing argument 2 of 'getsockname' from incompatible pointer type [-Werror=incompatible-pointer-types]
  902 |                 if (getsockname(this->send, &this->src, &addr_len) < 0)
      |                                             ^~~~~~~~~~
      |                                             |
      |                                             struct sockaddr_in *
In file included from /usr/include/fortify/sys/socket.h:23,
                 from /usr/include/netinet/in.h:10,
                 from /usr/include/arpa/inet.h:9,
                 from ../../../../src/libstrongswan/utils/utils.h:42,
                 from ../../../../src/libstrongswan/networking/host.h:28,
                 from dhcp_transaction.h:25,
                 from dhcp_socket.h:27,
                 from dhcp_socket.c:18:
/usr/include/sys/socket.h:391:23: note: expected 'struct sockaddr * restrict' but argument is of type 'struct sockaddr_in *'
  391 | int getsockname (int, struct sockaddr *__restrict, socklen_t *__restrict);
      |                       ^
cc1: all warnings being treated as errors
```

I will test the change with openresolv together with our VPN guys - they switched to 1 DNS server because of me, so I will ask them to try it again with 3 DNS servers 😅

**Reply by tobiasbrunner** [reply]
Pushed to the branch (seems like only newer compilers complain about this as none of our CI builds failed).

**Comment by oldium** [comment]
It works like a charm, tested with latest branch and `/etc/resolv.conf` writing works fine, same as with `resolvconf` feature. Thanks for good work and fast response!

**Reply by tobiasbrunner** [reply]
Great, thanks for testing. I'll line that up for the next release.
